File: app/src/main/java/com/wafflestudio/snutt_staging/view/TextRect.java
package com.wafflestudio.snutt_staging.view;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;

/**
 * Created by makesource on 2016. 1. 24..
 */
public class TextRect {
    // maximum number of lines; this is a fixed number in order
    // to use a predefined array to avoid ArrayList (or something
    // similar) because filling it does involve allocating memory
    static private int MAX_LINES = 256;

    // those members are stored per instance to minimize
    // the number of allocations to avoid triggering the
    // GC too much
    private Paint.FontMetricsInt metrics = null;
    private Paint paint = null;
    private int starts[] = new int[MAX_LINES];
    private int stops[] = new int[MAX_LINES];
    private int lines = 0;
    private int textHeight = 0;
    private Rect bounds = new Rect();
    private String text = null;
    private boolean wasCut = false;

    /**
     * Create reusable text rectangle (use one instance per font).
     *
     * @param paint - paint specifying the font
     */
    public TextRect( final Paint paint )
    {
        metrics = paint.getFontMetricsInt();
        String temp = metrics.toString();
        System.out.println("metrics" + temp);
        this.paint = paint;
    }

    /**
     * Calculate height of text block and prepare to draw it.
     *
     * @param text - text to draw
     * @param maxWidth - maximum width in pixels
     * @param maxHeight - maximum height in pixels
     * @returns height of text in pixels
     */
    public int prepare(final String text, final int maxWidth, final int maxHeight )
    {
        lines = 0;
        textHeight = 0;
        this.text = text;
        wasCut = false;

        // get maximum number of characters in one line
        paint.getTextBounds("i", 0, 1, bounds);

        final int maximumInLine = maxWidth / bounds.width();
        final int length = text.length();

        if( length > 0 )
        {
            final int lineHeight = -metrics.ascent + metrics.descent;
            int start = 0;
            int stop = maximumInLine > length ? length : maximumInLine;
            // stop = min(length,maximuminline);

            for( ;; )
            {
                // skip LF and spaces
                for( ; start < length; ++start )
                {
                    char ch = text.charAt( start );

                    if( ch != '\n' && ch != '\r' && ch != '\t' && ch != ' ' )
                        break;
                }

                for( int o = stop + 1; stop < o && stop > start; )
                {
                    o = stop;

                    int lowest = text.indexOf( "\n", start );
                    // 강의명만 생각

                    paint.getTextBounds(text, start, stop, bounds );

                    if( (lowest >= start && lowest < stop) ||
                            bounds.width() > maxWidth ) // start~stop까지 한 줄에 넣으려고 하는데 한 줄에 넘치면
                    {
                        --stop;

                        if( lowest < start ||
                                lowest > stop )
                        {
                            final int blank = text.lastIndexOf( " ", stop );
                            final int hyphen = text.lastIndexOf( "-", stop );

                            if( blank > start &&
                                    (hyphen < start || blank > hyphen) )
                                lowest = blank;
                            else if( hyphen > start )
                                lowest = hyphen;
                        }

                        if( lowest >= start &&
                                lowest <= stop )
                        {
                            final char ch = text.charAt( stop );

                            if( ch != '\n' &&
                                    ch != ' ' )
                                ++lowest;

                            stop = lowest;
                        }

                        continue;
                    }

                    break;
                }

                if( start >= stop )
                    break;

                int minus = 0;

                // cut off lf or space
                if( stop < length )
                {
                    final char ch = text.charAt( stop - 1 );

                    if( ch == '\n' ||
                            ch == ' ' )
                        minus = 1;
                }

                if( textHeight + lineHeight > maxHeight )
                {
                    wasCut = true;
                    break;
                }

                starts[lines] = start;
                stops[lines] = stop - minus;

                if( ++lines > MAX_LINES )
                {
                    wasCut = true;
                    break;
                }

                if( textHeight > 0 )
                    textHeight += metrics.leading;

                textHeight += lineHeight;

                if( stop >= length )
                    break;

                start = stop;
                stop = length;
            }
        }

        return textHeight;
    }

    /**
     * Draw prepared text at given position.
     *
     * @param canvas - canvas to draw text into
     * @param left - left corner
     * @param top - top corner
     */
    public void draw(final Canvas canvas, final int left, final int top, final int bubbleWidth, final int fgColor)
    {
        if( textHeight == 0 )
            return;

        final int before = -metrics.ascent;
        final int after = metrics.descent + metrics.leading;
        int y = top;

        --lines;
        for( int n = 0; n <= lines; ++n )
        {
            String t;

            y += before;

            if( wasCut && n == lines &&	stops[n] - starts[n] > 3 )
                t = text.substring( starts[n], stops[n] - 3 ).concat( "..." );
            else
                t = text.substring( starts[n], stops[n] );

            //텍스트 가운데 정렬
            int leftResult = left;
            paint.getTextBounds(t, 0, t.length(), bounds);
            int textWidth = bounds.right - bounds.left;
            leftResult += (bubbleWidth - textWidth) / 2;
            //
            paint.setColor(fgColor);
            canvas.drawText(t, leftResult, y, paint );

            y += after;
        }
    }

    /** Returns true if text was cut to fit into the maximum height */
    public final boolean wasCut()
    {
        return wasCut;
    }

}

File: app/src/main/java/com/wafflestudio/snutt2/components/view/TextRect.kt
package com.wafflestudio.snutt2.components.view

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Paint.FontMetricsInt
import android.graphics.Rect

class TextRect(private var paint: Paint) {
    private var metrics: FontMetricsInt = paint.fontMetricsInt
    private var singleLineHeight = 0
    private var cellWidth = 0
    private var lines = 0
    private var availableLines = 0
    private val starts = mutableListOf<Int>()
    private val stops = mutableListOf<Int>()
    private var textHeights = mutableListOf<Int>()
    private val bounds = Rect()
    private var text: String = ""
    private var wasCut = false
    private var toDraw = true

    fun prepare(text: String, cellWidth: Int, toDraw: Boolean) {
        clear()
        this.toDraw = toDraw
        if (text.isEmpty()) this.toDraw = false
        if (!this.toDraw) return
        this.text = text
        this.cellWidth = cellWidth
        cutToLines()
        availableLines = lines
    }

    fun draw(canvas: Canvas, left: Int, top: Int, fgColor: Int) {
        if (toDraw && availableLines > 0) {
            attachEllipsis()
            val before = -metrics.ascent
            val after = metrics.descent + metrics.leading
            var y = top
            for (n in 0 until availableLines) {
                y += before
                val textWithEllipsis = if (wasCut && n == availableLines - 1) {
                    text.substring(starts[n], stops[n]) + "..."
                } else {
                    text.substring(starts[n], stops[n])
                }

                // 텍스트 가운데 정렬
                var leftResult = left
                paint.getTextBounds(textWithEllipsis, 0, textWithEllipsis.length, bounds)
                leftResult += (cellWidth - (bounds.right - bounds.left)) / 2

                //
                paint.color = fgColor
                canvas.drawText(textWithEllipsis, leftResult.toFloat(), y.toFloat(), paint)
                y += after
            }
        }
    }

    fun getAvailableLines() = availableLines

    fun getTextHeight(reduceLine: Boolean = false): Int {
        if (!toDraw || availableLines == 0) return 0 // 항목이 공백으로만 되어있는 예외적인 경우 처리
        if (reduceLine) {
            availableLines -= 1
            wasCut = true
        }
        return textHeights[availableLines - 1]
    }

    fun getLeading(): Int = metrics.leading

    private fun clear() {
        lines = 0
        availableLines = 0
        starts.clear()
        stops.clear()
    }

    private fun cutToLines() {
        var textToPrepare = text
        while (textToPrepare.isNotEmpty()) {
            val textToCut = cutToSingleLine(textToPrepare)
            if (textToCut.isEmpty()) break // 끝이 공백으로 끝나는 예외적인 경우 처리

            starts.add(textToPrepare.indexOf(textToCut) + (if (lines == 0) 0 else stops[lines - 1]))
            stops.add(starts[lines] + textToCut.length)
            textHeights.add((if (lines == 0) 0 else textHeights[lines - 1]) + singleLineHeight)
            textToPrepare = text.substring(startIndex = stops[lines])
            lines += 1
        }
    }

    private fun cutToSingleLine(text: String): String {
        var result = ""
        for (char in text) {
            when {
                (char == '\n' && result.isNotEmpty()) -> break
                (char == '\n' && result.isEmpty()) -> continue
                (char == ' ' && result.isEmpty()) -> continue
            }

            result += char
            paint.getTextBounds(result, 0, result.length, bounds)
            if (bounds.width() > cellWidth) {
                while (result.lastIndexOf(' ') in 0 until result.length - 1) {
                    result = result.dropLast(1)
                }
                result = result.dropLast(1)
                break
            }
        }
        return result
    }

    private fun attachEllipsis() {
        if (wasCut) {
            while (true) {
                val lastLineText = text.substring(starts[availableLines - 1] until stops[availableLines - 1])
                paint.getTextBounds("$lastLineText...", 0, lastLineText.length + 3, bounds)
                if (bounds.width() <= cellWidth) break
                stops[availableLines - 1] -= 1
            }
        }
    }

    init {
        this.singleLineHeight = -metrics.ascent + metrics.descent + metrics.leading
    }
}
